### PR TITLE
feat(wallet): introduce Amount class for formatting numeric values

### PR DESCRIPTION
**Pull request by onyb**
## Motivation

Someone recently [pointed out](https://twitter.com/jim74255/status/1491103364746452994) a UX issue in our wallet with regards to how we display amounts.

## Previous work

[Fix inexact amounts in transction confirmation screen #11938](https://github.com/brave/brave-core/pull/11938).

## Proposed API

Instead of having tons of utility functions that are not flexible to different contexts, we now have a unified toolkit that allows conversion and formatting of amounts.

Here's what the API looks like to give a rough overview. We introduce an `Amount` class that wraps numeric values (hex, number, string, bignumber, etc). All relevant methods return new instances of `Amount`, in order to allow chaining of methods.

```ts
// Wei → ETH
new Amount(value)
  .divideByDecimals(18)

// ETH → Wei
new Amount(value)
  .multiplyByDecimals(18)

// Chain different stuff together.
new Amount(gasLimit)
  .times(gasPrice)
  .divideByDecimals(18)
  .times(usdPriceOfEth)
  .format()

// Formatting batteries included.
new Amount(whateverAmount)
  .format(6, true) // 6 significant figures, formatted with commas
new Amount(money)
  .formatAsFiat(currency) // use fiat formatting rules, prefixed by currency symbol
```

## Improved precision

We now are able to format amounts with much better precision, thanks to **use of significant figures** as opposed to rounding to a fixed number of decimal places.

If you are interested in the concept of significant figures and the underlying rules, please refer to [this Khan Academy video](https://www.khanacademy.org/math/arithmetic-home/arith-review-decimals/arithmetic-significant-figures-tutorial/v/significant-figures) which I found very helpful.

⚠️ Use of significant figures is avoided when values are too large to get any benefit from such formatting. In such cases, we fallback to formatting to a fixed number of decimals.

For further information, you can also refer to the Demo section. PR includes unit-tests with 💯 coverage, which is a good starting point for code reviewers.

Resolves https://github.com/brave/brave-browser/issues/20994.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Demo

🔍  Zoom into each image for further information.

| Portfolio | TXN history | Fees popup | TXN confirmation | EIP-1559 |
-|-|-|-|-
| ||||  |

### CODE REVIEW

File: components/brave_wallet_ui/utils/amount.ts [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    }
+
+    if (value === '' || this.value === undefined) {
+      return Amount.empty()
+    }
+
+    return new Amount(this.value.times(value))
+  }
+
+  div (value: AmountLike): Amount {
```
Comment (muliswilliam) [inline components/brave_wallet_ui/utils/amount.ts:57]: Can we add a check if the value is 0 to avoid getting `Infinity` as a result of division by 0?

**Review by Douglashdaniel — APPROVED** [review]
Solid!
Tested locally as well, looks great!!

File: components/brave_wallet_ui/components/desktop/portfolio-transaction-item/index.tsx [hunk 2]
```diff
@@ -262,6 +258,8 @@ const PortfolioTransactionItem = (props: Props) => {
     }
   }, [transactionDetails])
 
+  console.log(transactionDetails)
```
[the remark this answers is not in the document]
Comment (onyb) [reply components/brave_wallet_ui/components/desktop/portfolio-transaction-item/index.tsx:261]: Done

File: components/brave_wallet_ui/utils/amount.ts (hunk 1 above)
Comment (onyb) [reply components/brave_wallet_ui/utils/amount.ts:57]: Fixed and added a test.

**Pull request merged**